File: Java/811.java
__________________________________________________________________________________________________
sample 5 ms submission
class Solution {
    public List<String> subdomainVisits(String[] cpdomains) {
        Map<String, DomainInfo> cmap = new HashMap<>();
        for (String cpdomain : cpdomains) {
            addup(cmap, cpdomain);
        }
        ArrayList<String> result = new ArrayList<>(cmap.size());
        for (Map.Entry<String, DomainInfo> e : cmap.entrySet()) {
            result.add(e.getValue().count + " " + e.getKey()); 
        }
        return result;
    }
    
    void addup(Map<String, DomainInfo> cmap, String cpdomain) {
        int spaceIndex = cpdomain.indexOf(' ');
        int count = Integer.parseInt(cpdomain.substring(0, spaceIndex));
        String domain = cpdomain.substring(spaceIndex + 1);
        addInfo(cmap.get(domain), cmap, count, domain);
    }
    
    DomainInfo addInfo(DomainInfo d, Map<String, DomainInfo> cmap, int count, 
                       String domain) {
        if (d == null) {
            // there is nothing to do if the domain is null too
            if (domain == null) {
                return null;
            }
            final String updomain = getUpDomain(domain);
            d = new DomainInfo(count, addInfo(cmap.get(updomain), cmap, count, updomain));
            cmap.put(domain, d);
        } else {
            d.count += count;
            for (DomainInfo tmp = d.updomain; tmp != null; tmp = tmp.updomain) {
                tmp.count += count;
            }
        }
        return d;
    }
                 
    String getUpDomain(String domain) {
        int dotIndex = domain.indexOf('.');
        if (dotIndex >= 0) {
            return domain.substring(dotIndex + 1);
        }
        return null;
    }
    
    static class DomainInfo {
        int count;
        DomainInfo updomain;
        
        DomainInfo(int count, DomainInfo updomain) {
            this.count = count;
            this.updomain = updomain;
        }
    }
}
__________________________________________________________________________________________________
sample 35976 kb submission
class Solution {
    public List<String> subdomainVisits(String[] cpdomains) {
        final Map<String, Integer> count = new HashMap<>();
        for(String s : cpdomains) {
            String ss[] = s.split(" ");
            int num = Integer.parseInt(ss[0]);
            String domain = ss[1];
            ss = domain.split("\\.");
            count.put(
                domain,
                count.computeIfAbsent(domain, k->0) + num
            );
            int sum = 1;
            for(int i = 0; i < ss.length - 1; i++) {
                sum += ss[i].length() + i;
                String sub = domain.substring(sum);
                count.put(
                    sub,
                    count.computeIfAbsent(sub, k->0) + num
               );
            }
        }
        final List<String> all = new ArrayList<>();
        for(Map.Entry<String, Integer> e : count.entrySet())
            all.add(
                new StringBuilder().append(e.getValue())
                                   .append(" ")
                                   .append(e.getKey())
                                   .toString()
            );
        return all;
    }
}
__________________________________________________________________________________________________

File: Python3/811.py
__________________________________________________________________________________________________
sample 44 ms submission
class Solution:
    def subdomainVisits(self, cpdomains: List[str]) -> List[str]:
        tempDict = {}
        for i in cpdomains:
            count,domain = i.split(' ')
            if domain not in tempDict:
                tempDict[domain] = int(count)
            else:
                tempDict[domain] += int(count)
            
            subdomain = domain.split('.')[1::]
            temp = ''
            for j in range(len(subdomain)-1, -1, -1):
                if temp == '':
                    temp = subdomain[j]
                else:
                    temp = subdomain[j] + '.' + temp
                if temp not in tempDict:
                    tempDict[temp] = int(count)
                else:
                    tempDict[temp] += int(count)
        result = []
        for k,v in tempDict.items():
            result.append(str(v) + ' ' + k)
        return result
__________________________________________________________________________________________________
sample 13036 kb submission
class Solution:
    def subdomainVisits(self, cpdomains: List[str]) -> List[str]:
        
        dic={}
        for i in cpdomains:
            temp = i.split(" ")
            
            # temp[0] is the count 
            # temp[1] is the domain name 
            
            subdomain = []
            subdomain.append(temp[1])
            a= temp[1]
            b=temp[1]
            for k in range(0,a.count(".")):
                index = b.find(".")
                string = b[index+1:]
                subdomain.append(string)
                b = string
        
            count= int(temp[0])
            for j in subdomain:
                if j in dic:
                    dic[j] = dic[j]+count
                else:
                    dic[j] = count 
        
        

        result = [] 
        for i in range(0,len(dic)):
            key, value = dic.popitem()
            string = str(value)+" " + key
            result.append(string)
        
        return result
__________________________________________________________________________________________________
